<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>zinLDA: a vignette • zinLDA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="zinLDA: a vignette">
<meta property="og:description" content="zinLDA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">zinLDA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/zinLDA-vignette.html">zinLDA: a vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="zinLDA-vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>zinLDA: a vignette</h1>
                        <h4 class="author">Rebecca A. Deek</h4>
            
      
      
      <div class="hidden name"><code>zinLDA-vignette.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">zinLDA</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://purrr.tidyverse.org">purrr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape">reshape2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/juliasilge/tidytext">tidytext</a></span><span class="op">)</span></pre></div>
<div id="about" class="section level2">
<h2 class="hasAnchor">
<a href="#about" class="anchor"></a>About</h2>
<p>The <code>zinLDA</code> package provides a Markov chain Monte Carlo implementation of a Zero-Inflated Latent Dirichlet Allocation (zinLDA) model. It also provides a simulation algorithm for generating sparse count data from an underlying zinLDA model.</p>
</div>
<div id="what-is-zero-inflated-latent-dirichlet-allocation" class="section level2">
<h2 class="hasAnchor">
<a href="#what-is-zero-inflated-latent-dirichlet-allocation" class="anchor"></a>What is Zero-Inflated Latent Dirichlet Allocation?</h2>
<p>Zero-inflated latent Dirichlet allocation is an unsupervised, hierarchical, generative probabilistic model. It is flexible enough to be applied to count data in a wide range of fields from text analysis to bioinformatics. In particular, when applied to microbial 16S rRNA gene sequencing studies, which are notoriously high dimensional and sparse, it is a powerful method for dimensionality reduction and detection of potentially biologically meaningful sparse clusters of taxa. These clusters of taxa are called subcommunities here.</p>
<p>The zinLDA model assumes that a single microbial sample contains <span class="math inline">\(K\)</span> different underlying subcommunities with varying probability. In terms of the generative model, this means that zinLDA samples the latent subcommunity indicator once per 16S sequencing read, as compared to the Dirichlet-multinomial mixture model, commonly used in microbiome studies, that assumes the subcommunity label is the same across all sequencing reads within a microbial sample. Furthermore, zinLDA assumes that each taxon can be assigned a probability of belonging to each of the <span class="math inline">\(K\)</span> subcommunities.</p>
<p>In particular, zinLDA adapts the original Latent Dirichlet Allocation (LDA) model of Blei, Ng, and Jordan (2003) to better handle sparse count data by assuming that these taxa per subcommunity probabilities are sparse, meaning that some, if not many, of them are equal to zero. By doing so, within each subcommunity, the zinLDA model differentiates structural zeros, taxa that truly do not belong to the community, from sampling zeros, taxa that do contribute to the community but due to low sequencing depth were not captured.</p>
<div id="notation-and-terminology" class="section level3">
<h3 class="hasAnchor">
<a href="#notation-and-terminology" class="anchor"></a>Notation and Terminology</h3>
<p>Before giving the model specification for zinLDA, it is important to define the notation of key variables and model parameters that will be used for the remainder of this vignette.</p>
<p>For a single 16S rRNA gene sequencing sample define:</p>
<ul>
<li>
<span class="math inline">\(w_{dn}\)</span> as the <span class="math inline">\(n^{th}\)</span> observed sequencing read in the <span class="math inline">\(d^{th}\)</span> biological sample. It is represented by <span class="math inline">\(V\)</span>-length vector with a single non-zero component, which takes values equal to one.</li>
<li>
<span class="math inline">\(V\)</span> as the number of unique taxa in the study.</li>
<li>
<span class="math inline">\(w_{dn}^{i}\)</span> as the <span class="math inline">\(n^{th}\)</span> sequencing read in the <span class="math inline">\(d^{th}\)</span> sample which assigned to the <span class="math inline">\(i^{th}\)</span> unique taxa.</li>
<li>
<span class="math inline">\(\boldsymbol{w}_{d} = (w_{d1}, \dots, w_{dN})\)</span> as the <span class="math inline">\(d^{th}\)</span> biological sample consisting of <span class="math inline">\(N\)</span> sequencing reads.</li>
<li>
<span class="math inline">\(\textbf{D}\)</span> = (<span class="math inline">\(\boldsymbol{w}_1, \dots, \boldsymbol{w}_{D}\)</span>) as the collection of all biological samples in the study, called a cohort.</li>
<li>
<span class="math inline">\(z_{dn} = j\)</span> as the indicator that the <span class="math inline">\(n^{th}\)</span> sequencing read in the <span class="math inline">\(d^{th}\)</span> sample belongs to the <span class="math inline">\(j^{th}\)</span> subcommunity.</li>
<li>
<span class="math inline">\(\boldsymbol{z}^{(d)} = (z_{d1}, \dots, z_{dN})\)</span> as the vector of subcommunity assignments for the <span class="math inline">\(d^{th}\)</span> biological sample.</li>
</ul>
</div>
<div id="model-specification" class="section level3">
<h3 class="hasAnchor">
<a href="#model-specification" class="anchor"></a>Model Specification</h3>
<p>Now a generative probabilistic model for zinLDA can be defined as follows:</p>
<ol style="list-style-type: decimal">
<li>For each of the <span class="math inline">\(K\)</span> subcommunities, indexed by <span class="math inline">\(j\)</span>:
<ul>
<li>Choose <span class="math inline">\(\boldsymbol{\Delta}^{(j)} \sim\)</span> Ber(<span class="math inline">\(\pi\)</span>), where <span class="math inline">\(\Delta_{ij} = I(\beta_{ij} = 0)\)</span> is an indicator variable for structural zeros.</li>
<li>Choose <span class="math inline">\(\boldsymbol{\beta}^{(j)} \sim\)</span> ZIGD(<span class="math inline">\(\pi, a, b\)</span>), where <span class="math inline">\(\beta_{ij}\)</span> = Pr<span class="math inline">\((w^{i} \vert z=j)\)</span> is the probability that the <span class="math inline">\(i^{th}\)</span> taxon belongs to the <span class="math inline">\(j^{th}\)</span> subcommunity.</li>
</ul>
</li>
<li>For each biological sample <span class="math inline">\(\boldsymbol{w_{d}}\)</span> in the cohort:
<ul>
<li>Choose <span class="math inline">\(\boldsymbol{\theta^{(d)}} \sim\)</span> Dir(<span class="math inline">\(\alpha\)</span>), where <span class="math inline">\(\theta_{dj}\)</span> = Pr<span class="math inline">\((z_{d}=j)\)</span> is the probability of observing the <span class="math inline">\(j^{th}\)</span> subcommunity in the <span class="math inline">\(d^{th}\)</span> sample.</li>
</ul>
</li>
<li>For each of the N sequencing reads, <span class="math inline">\(w_{dn}\)</span>:
<ul>
<li>Choose a subcommunity, <span class="math inline">\(z_{dn} \sim\)</span> Multinomial(1, <span class="math inline">\(\boldsymbol{\theta}^{(d)}\)</span>).</li>
<li>Choose a taxon, <span class="math inline">\(w_{dn}\)</span> from Pr<span class="math inline">\((w_{dn} \vert z_{dn}, \boldsymbol{\beta})\)</span>, a multinomial probability distribution conditional on the subcommunity <span class="math inline">\(z_{dn}\)</span>.</li>
</ul>
</li>
</ol>
<p>The key difference of this generative model and that of LDA is the addition of the Zero-Inflated Generalized Dirichlet (ZIGD) distribution on <span class="math inline">\(\boldsymbol{\beta}^{(j)}\)</span>, the taxa distribution for subcommunity <span class="math inline">\(j\)</span>. The ZIGD distribution is parameterized by <span class="math inline">\(\pi, a,\)</span> and <span class="math inline">\(b\)</span>. Here, <span class="math inline">\(\pi\)</span> is the probability of being a structural zero, while <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> are the parameters of a generalized Dirichlet distribution, an extension of the Dirichlet distribution with a more flexible covariance structure. The additional covariance parameters enable ZIGD to better capture the complex dynamics and heterogeneity between microbial taxa. Additionally, this distributional specification is what allows zinLDA to capture both sampling and structural zeros.</p>
</div>
</div>
<div id="simulating-data" class="section level2">
<h2 class="hasAnchor">
<a href="#simulating-data" class="anchor"></a>Simulating Data</h2>
<p>Below is an outline on how to simulate realistic microbiome data from a zero-inflated latent Dirichlet allocation model using the <code><a href="../reference/simulateZINLDA.html">simulateZINLDA()</a></code> function. First, let’s assume the cohort of interest consists of 40 microbial samples and that the total number of sequencing reads per sample is drawn from a discrete uniform distribution with lower bound 100 and upper bound 500. It should be noted that the sequencing depth of a typical 16S rRNA gene sequencing microbial sample is on the order of thousands or tens of thousands of reads per sample. The small number of sequences reads was selected to ease computation time.</p>
<p>Moreover, let’s assume that there are 50 unique microbial taxa that are present across the 40 subjects, each with a probability of belonging to the four underlying subcommunity structures. For hyperparameter specification, <span class="math inline">\(\alpha\)</span>, the parameter indexing the Dirichlet distribution on <span class="math inline">\(\theta\)</span> is set to a default value of 0.1, as recommended in the Gibbs sampling implementation of the original LDA model. The value of <span class="math inline">\(\pi\)</span>, <span class="math inline">\(a\)</span>, and <span class="math inline">\(b\)</span> are set to 0.4, 0.5, and 10, respectively.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">N.d</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/rdu.html">rdu</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">40</span>, min <span class="op">=</span> <span class="fl">100</span>, max <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>

<span class="va">sim1</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/simulateZINLDA.html">simulateZINLDA</a></span><span class="op">(</span>D <span class="op">=</span> <span class="fl">40</span>, V <span class="op">=</span> <span class="fl">50</span>, N <span class="op">=</span> <span class="va">N.d</span>, K <span class="op">=</span> <span class="fl">4</span>, 
                              Alpha <span class="op">=</span> <span class="fl">0.1</span>, Pi <span class="op">=</span> <span class="fl">0.4</span>, a <span class="op">=</span> <span class="fl">.5</span>, b <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></pre></div>
<p>The <code>simulateZINLDA</code> function returns a list with six elements. The first, <code>cohort</code>, is a D-length list containing the taxonomic assignment of each sequencing reads in the D samples. Below is the first six taxa assignments of the first sample:</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sim1</span><span class="op">$</span><span class="va">cohort</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; [1] "Taxa 43" "Taxa 13" "Taxa 13" "Taxa 16" "Taxa 14" "Taxa 49"</span></pre></div>
<p>Second is <code>z</code>, another D-length list containing the subcommunity indicator for each of the taxonomically assigned sequencing reads in <code>cohort</code>. Below shows the subcommunity assignments for the first six sequencing reads of the first sample:</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sim1</span><span class="op">$</span><span class="va">z</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; Taxa 43 Taxa 13 Taxa 13 Taxa 16 Taxa 14 Taxa 49 </span>
<span class="co">#&gt;       3       3       3       3       3       3</span></pre></div>
<p>Next, <code>sampleTaxaMatrix</code> is a <span class="math inline">\(D \times V\)</span> matrix containing the count data of interest. Each row of the matrix can be thought of as a V-length vector of taxon counts for a single sample. Below are the counts of the first five taxa in the first five samples:</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">sim1</span><span class="op">$</span><span class="va">sampleTaxaMatrix</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;          Taxa 1 Taxa 2 Taxa 3 Taxa 4 Taxa 5</span>
<span class="co">#&gt; Sample 1      2      1      2      8      0</span>
<span class="co">#&gt; Sample 2      0      7     29      1      0</span>
<span class="co">#&gt; Sample 3      0      0      0      0      4</span>
<span class="co">#&gt; Sample 4      1      2      0      2      0</span>
<span class="co">#&gt; Sample 5      0      2      0      7      3</span></pre></div>
<p>The last three elements are three matrices: <span class="math inline">\(D \times K\)</span> matrix <code>theta</code>, <span class="math inline">\(K \times V\)</span> matrix <code>beta</code>, and <span class="math inline">\(K \times V\)</span> matrix <code>delta</code>. The first two contain the subcommunity per sample probabilities and the taxa per subcommunity probabilities, respectively. While <code>delta</code> is a matrix of 0s and 1s identifying which taxa are sampling zeros in each subcommunity. Below are the first few entries of each:</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="va">sim1</span><span class="op">$</span><span class="va">theta</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span>
<span class="co">#&gt;           Community 1 Community 2  Community 3  Community 4</span>
<span class="co">#&gt; Sample 1 6.783018e-03 0.002874889 9.077920e-01 8.255006e-02</span>
<span class="co">#&gt; Sample 2 9.678068e-01 0.020301022 1.189214e-02 5.524276e-13</span>
<span class="co">#&gt; Sample 3 3.478782e-05 0.985956309 1.269506e-13 1.400890e-02</span>
<span class="co">#&gt; Sample 4 1.070988e-04 0.003240278 9.825775e-01 1.407510e-02</span>
<span class="co">#&gt; Sample 5 7.562609e-04 0.251068287 7.481754e-01 9.948178e-08</span>
<span class="va">sim1</span><span class="op">$</span><span class="va">beta</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;                  Taxa 1       Taxa 2       Taxa 3      Taxa 4     Taxa 5</span>
<span class="co">#&gt; Community 1 0.000000000 4.816090e-02 0.1791166975 0.004059997 0.00000000</span>
<span class="co">#&gt; Community 2 0.002044051 7.237084e-05 0.0000000000 0.000000000 0.04503479</span>
<span class="co">#&gt; Community 3 0.010152121 1.258581e-02 0.0000482654 0.036234391 0.00000000</span>
<span class="co">#&gt; Community 4 0.035341655 0.000000e+00 0.0934088401 0.015919066 0.01391503</span>
<span class="va">sim1</span><span class="op">$</span><span class="va">delta</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;             Taxa 1 Taxa 2 Taxa 3 Taxa 4 Taxa 5</span>
<span class="co">#&gt; Community 1      1      0      0      0      1</span>
<span class="co">#&gt; Community 2      0      0      1      1      0</span>
<span class="co">#&gt; Community 3      0      0      0      0      1</span>
<span class="co">#&gt; Community 4      0      1      0      0      0</span></pre></div>
<p>Since microbiome data is typically very sparse let’s calculate the sparsity, defined as the percent of zeros, of the simulated count data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">sparsity</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sim</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="va">sampleTaxaMatrix</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/prod.html">prod</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="va">sampleTaxaMatrix</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">sparsity</span><span class="op">(</span><span class="va">sim1</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.489</span></pre></div>
<p>The sparsity of the taxon count matrix is 48.9%, which is on par with that of species count matrices filtered to a prevalence of 20% from large microbiome studies such as the American Gut Project. It is helpful to visualize the sparsity of a matrix using a heatmap. Below is a heatmap of the arcsinh-transformed taxon counts from simulation, the y-axis lists the samples and the x-axis the taxa. The white represents a count of zero while the darker the blue indicates a larger count.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">abunHeatmap</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">sim</span><span class="op">)</span><span class="op">{</span>
  <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Hyperbolic.html">asinh</a></span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="va">sampleTaxaMatrix</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="va">Var1</span>, taxa <span class="op">=</span> <span class="va">Var2</span>, asinh.abun <span class="op">=</span> <span class="va">value</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span> <span class="op">(</span>x <span class="op">=</span> <span class="va">taxa</span>, y <span class="op">=</span> <span class="va">sample</span>, fill <span class="op">=</span> <span class="va">asinh.abun</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"arcsinh\nabundance"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,
        axis.title.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.title.y<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">3</span>, angle<span class="op">=</span><span class="fl">90</span><span class="op">)</span>, axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient</a></span><span class="op">(</span>low<span class="op">=</span><span class="st">"white"</span>, high<span class="op">=</span><span class="st">"#2171b5"</span>, na.value <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">abunHeatmap</span><span class="op">(</span><span class="va">sim1</span><span class="op">)</span></pre></div>
<p><img src="zinLDA-vignette_files/figure-html/heatmap-1-1.png" width="480"></p>
<p>Below the hyperparameter values are varied in different simulations to demonstrate how hyperparameter choice influences the sparsity of the resulting taxon counts. Sparsity is strongly influences by the probability of being a structural or sampling zero. It is expected that increasing the probability of being a sampling zero increases the sparsity, while shrinking some of the <span class="math inline">\(\beta_{ij}\)</span> towards 0 will increasing the number of sampling zeros and therefore the overall sparsity.</p>
<p>For example, changing the values of <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> in <code>sim2</code> increases the median value of <span class="math inline">\(\beta_{ij}\)</span> meaning that the probability of observing a structural zero decreases, thus in turn decreasing the overall sparsity of the matrix. Conversely, decreasing <span class="math inline">\(V\)</span> in <code>sim3</code>, while holding all other parameters values constant, decreases the median of <span class="math inline">\(\beta_{ij}\)</span> and increases both the probability of being a sampling zero and the sparsity of the matrix. Lastly, increasing <span class="math inline">\(\pi\)</span> in <code>sim4</code> increases the probability of being a structural zero and the sparsity of the matrix.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">sim2</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/simulateZINLDA.html">simulateZINLDA</a></span><span class="op">(</span>D <span class="op">=</span> <span class="fl">40</span>, V <span class="op">=</span> <span class="fl">50</span>, N <span class="op">=</span> <span class="va">N.d</span>, K <span class="op">=</span> <span class="fl">4</span>, 
                              Alpha <span class="op">=</span> <span class="fl">0.1</span>, Pi <span class="op">=</span> <span class="fl">0.4</span>, a <span class="op">=</span> <span class="fl">1</span>, b <span class="op">=</span> <span class="fl">25</span><span class="op">)</span>

<span class="va">sim3</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/simulateZINLDA.html">simulateZINLDA</a></span><span class="op">(</span>D <span class="op">=</span> <span class="fl">40</span>, V <span class="op">=</span> <span class="fl">25</span>, N <span class="op">=</span> <span class="va">N.d</span>, K <span class="op">=</span> <span class="fl">4</span>, 
                              Alpha <span class="op">=</span> <span class="fl">0.1</span>, Pi <span class="op">=</span> <span class="fl">0.4</span>, a <span class="op">=</span> <span class="fl">.5</span>, b <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>

<span class="va">sim4</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/simulateZINLDA.html">simulateZINLDA</a></span><span class="op">(</span>D <span class="op">=</span> <span class="fl">40</span>, V <span class="op">=</span> <span class="fl">50</span>, N <span class="op">=</span> <span class="va">N.d</span>, K <span class="op">=</span> <span class="fl">4</span>, 
                              Alpha <span class="op">=</span> <span class="fl">0.1</span>, Pi <span class="op">=</span> <span class="fl">0.7</span>, a <span class="op">=</span> <span class="fl">.5</span>, b <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>

<span class="va">sim.ls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">sim2</span>,<span class="va">sim3</span>,<span class="va">sim4</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">sim.ls</span>, <span class="va">sparsity</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; [1] 0.394</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; [1] 0.517</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; [1] 0.639</span>

<span class="va">asinh.abun.hmp</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html">map2</a></span><span class="op">(</span><span class="va">sim.ls</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">4</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span> <span class="op">{</span> 
                                            <span class="fu">abunHeatmap</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">+</span>
                                              <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span> <span class="op">}</span><span class="op">)</span>
<span class="va">count.hmp.ls</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"grid.arrange"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">asinh.abun.hmp</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></pre></div>
<p><img src="zinLDA-vignette_files/figure-html/sparsity-ls-1.png" width="672"></p>
<p>Another advantageous feature of using zinLDA to simulate realistic microbiome data is that it can provide some information about the number of taxa that are missed due to low sequencing depth. For example, <code>sim1</code> and <code>sim4</code> both contain taxa that have zero counts for every sample in the cohort, thus in a real microbiome data set these taxa would not be observed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">sim1</span>,<span class="va">sim2</span>,<span class="va">sim3</span>,<span class="va">sim4</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">sampleTaxaMatrix</span><span class="op">)</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; Taxa 25 Taxa 47 </span>
<span class="co">#&gt;      25      47 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; named integer(0)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; named integer(0)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[4]]</span>
<span class="co">#&gt; Taxa 16 Taxa 21 Taxa 33 Taxa 43 Taxa 44 </span>
<span class="co">#&gt;      16      21      33      43      44</span></pre></div>
</div>
<div id="model-fitting" class="section level2">
<h2 class="hasAnchor">
<a href="#model-fitting" class="anchor"></a>Model Fitting</h2>
<p>Now let’s assume this data was not simulated and the true values of <span class="math inline">\(\boldsymbol\beta\)</span> and <span class="math inline">\(\boldsymbol\theta\)</span> are unknown. A zinLDA model can be fit to the data to estimate them. Below, we set the number of latent communities and the hyperparameters to their true values as specified under simulation. Though, in general, picking the number of subcommunities, or clusters, is a difficult question to answer. Ideally, biological or clinical reasoning would inform the choice of <span class="math inline">\(K\)</span>, if available. Otherwise, data-driven approaches, such as fitting multiple zinLDA models to a subset of the data and selecting the one that gives the smallest log-likelihood or AIC, can be used.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">modelFit1</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/zinLDA.html">zinLDA</a></span><span class="op">(</span><span class="va">sim1</span><span class="op">$</span><span class="va">sampleTaxaMatrix</span>, K <span class="op">=</span> <span class="fl">4</span>, alpha <span class="op">=</span> <span class="fl">.1</span>, pi <span class="op">=</span> <span class="fl">0.4</span>, a <span class="op">=</span> <span class="fl">.5</span>, b <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">posteriorEsts</span> <span class="op">=</span> <span class="fu">zinLDA</span><span class="fu">::</span><span class="fu"><a href="../reference/posterior.html">posterior</a></span><span class="op">(</span><span class="va">modelFit1</span><span class="op">)</span></pre></div>
<p>Any given subcommunity can be described by its representative taxa, defined as the taxa with the highest probability of belonging to that community. Here the five taxa with the the highest probability, <span class="math inline">\(\beta_{ij}\)</span>, for each of the four subcommunities are plotted.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html">melt</a></span><span class="op">(</span><span class="va">posteriorEsts</span><span class="op">$</span><span class="va">beta</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>community<span class="op">=</span><span class="va">Var1</span>, taxa<span class="op">=</span><span class="va">Var2</span>, beta<span class="op">=</span><span class="va">value</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">community</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span><span class="fl">5</span>, <span class="va">beta</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">community</span>, <span class="op">-</span><span class="va">beta</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>taxa <span class="op">=</span> <span class="fu">tidytext</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidytext/man/reorder_within.html">reorder_within</a></span><span class="op">(</span><span class="va">taxa</span>, <span class="va">beta</span>, <span class="va">community</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">taxa</span>, <span class="va">beta</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">community</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span>show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">community</span>, scales <span class="op">=</span> <span class="st">"free"</span>, nrow <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">tidytext</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidytext/man/reorder_within.html">scale_x_reordered</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Top Taxa per Subcommunity"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html">expression</a></span><span class="op">(</span><span class="va">beta</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>, axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">beta</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="op">-</span><span class="fl">.1</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span></pre></div>
<p><img src="zinLDA-vignette_files/figure-html/top-taxa-1.png" width="672"></p>
<p>Other downstream analyses that may be of interest are using the using the subcommunity per sample distributions in a compositional regression to identify clinical covariates associated with subcommunity types or using the subcommunity per sample distributions to identify population-level clusters of subjects using hierarchical or k-means clustering.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<ol style="list-style-type: decimal">
<li>
<a href="https://www.jmlr.org/papers/volume3/blei03a/blei03a.pdf">Latent dirichlet allocation</a> Blei, D. M., Ng, A. Y., &amp; Jordan, M. I. (2003).</li>
<li>
<a href="https://www.pnas.org/content/pnas/101/suppl_1/5228.full.pdf">Finding scientific topics</a> Griffiths, T. L., &amp; Steyvers, M. (2004).</li>
<li>
<a href="https://academic.oup.com/biostatistics/article-abstract/20/4/698/5043450">Zero-inflated generalized Dirichlet multinomial regression model for microbiome compositional data analysis</a> Tang, Z. Z., &amp; Chen, G. (2019).</li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Rebecca A. Deek.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
